### PR TITLE
On dotcom snippetAttribution resolver performs search.

### LINKED ISSUE #59694: Attribution endpoint broken
https://sourcegraph.slack.com/archives/C05497E9MDW/p1705574718346319

### PR DESCRIPTION
**Pull request by cbart**
Closes #59694

in https://github.com/sourcegraph/sourcegraph/pull/59513 I broke `snippetAttribution` on dotcom. We're bringing back local (on dotcom) snippet attribution which is backed by search. On an enterprise instance `snippetAttribution` still federates to gateway and ends up in a dotcom search (via the same graphQL op on doctom).

## Test plan

- Unit test with fake search client.
- Manual check.

### CODE REVIEW

File: cmd/frontend/internal/guardrails/attribution/attribution.go [hunk 1]
```diff
@@ -75,3 +88,92 @@ func (c *Service) SnippetAttribution(ctx context.Context, snippet string, limit
...
+// Note: this registers metrics so should only be called once with the same
+// observationCtx.
+func NewLocalSearch(observationCtx *observation.Context, client client.SearchClient) Service {
+	return &localSearch{
+		operations: newOperations(observationCtx),
+		client:     client,
+	}
+}
+
+func (c *localSearch) SnippetAttribution(ctx context.Context, snippet string, limit int) (result *SnippetAttributions, err error) {
```
Comment (cbart) [inline cmd/frontend/internal/guardrails/attribution/attribution.go:108]: Whole method here is taken verbatim from the previous implementation of the resolver: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f3f8151c51495803d35539fc883d3abd290f007b/-/blob/cmd/frontend/internal/guardrails/attribution/attribution.go?L169

File: cmd/frontend/internal/guardrails/attribution/attribution.go [hunk 2]
```diff
@@ -2,31 +2,27 @@ package attribution
...
+	"github.com/sourcegraph/sourcegraph/internal/search/client"
+	"github.com/sourcegraph/sourcegraph/internal/search/streaming"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
 )
 
 // Service is for the attribution service which searches for matches on
 // snippets of code.
 //
 // Use NewService to construct this value.
```
Comment (slimsag) [inline cmd/frontend/internal/guardrails/attribution/attribution.go:23]:
```suggestion
```

there is no NewService

File: cmd/frontend/internal/guardrails/attribution/attribution.go [hunk 3]
```diff
@@ -55,9 +51,26 @@ type SnippetAttributions struct {
 	LimitHit bool
 }
 
+// gatewayProxy is a Service that proxies requests to cody gateway.
+type gatewayProxy struct {
+	client     codygateway.Client
+	operations *operations
+}
+
+// NewGatewayProxy returns an attribution service that proxies to a gateway request.
```
Comment (slimsag) [inline cmd/frontend/internal/guardrails/attribution/attribution.go:60]: What is the architecture here? I guess the editor hits a Sourcegraph instance first, then the instance decies whether to redirect to local search or proxy to the gateway, but then what does the gateway do? It proxies to dotcom? Or it has its own attribution search somehow?

**Review by slimsag — APPROVED** [review]

File: cmd/frontend/internal/guardrails/attribution/attribution.go (hunk 3 above)
Comment (cbart) [reply cmd/frontend/internal/guardrails/attribution/attribution.go:59]: It's just one:

Editor -> SG instance -> gateway -> dotcom (peroforms search)

SG instance and dotcom serve on the same graphQL, hence the dispatch here.

The reason for hops is authentication:

- Editor authenticates with user token 
- Enterprise instance authenticates with product subscription
- Gateway is hardwired with dotcom

**Pull request merged**